### ISSUE TITLE
Unable to run any Libraries tests on Ubuntu 22.04 ARM64

### ISSUE THREAD

**Issue by vcsjones** [dotnet/runtime#75374]
Using a brand new installation of Ubuntu 22.04 ARM64, I am unable to execute any tests for `dotnet/runtime`. It always results in "Testiest process exited with error: . Please check the diagnostic logs for more information".

Steps to reproduce.

1. Fresh install of Ubuntu 22.04 ARM64
2. Install dependencies as documented in the [workflow instructions](https://github.com/dotnet/runtime/blob/main/docs/workflow/requirements/linux-requirements.md)
    Note: I needed to substitute `python` for `python3`.
3. Run the build: `./build.sh -rc release -s clr+libs`
4. Try to run a test: `/dotnet.sh test src/libraries/System.Formats.Asn1/tests`.
    Note: I tried other projects as well like `System.Security.Cryptography` and `System.Formats.Tar`, all have the same result.

Observe the crash.

I can only reproduce this on ARM64, not AMD64.

**Comment by vcsjones** [comment]
Using `testhost` directly, this is what LLDB shows:

```plain
* thread #1, name = 'dotnet', stop reason = signal SIGSEGV: address access protected (fault address: 0xffffa826a288)
  * frame #0: 0x0000ffffa824a288
    frame #1: 0x0000fffff747525c libcoreclr.so`MethodDescCallSite::CallTargetWorker(unsigned long const*, unsigned long*, int) at callhelpers.cpp:67:5
    frame #2: 0x0000fffff7475204 libcoreclr.so`MethodDescCallSite::CallTargetWorker(this=<unavailable>, pArguments=<unavailable>, pReturnValue=0x0000000000000000, cbReturnValue=0) at callhelpers.cpp:562:9
    frame #3: 0x0000fffff7399ac0 libcoreclr.so`CorHost2::CreateAppDomainWithManager(char16_t const*, unsigned int, char16_t const*, char16_t const*, int, char16_t const**, char16_t const**, unsigned int*) [inlined] MethodDescCallSite::Call(this=0x0000ffffffffe4b8, pArguments=0x0000ffffffffe5d8) at callhelpers.h:458:9
    frame #4: 0x0000fffff7399aac libcoreclr.so`CorHost2::CreateAppDomainWithManager(this=0x0000aaaaaab3cfa0, wszFriendlyName=<unavailable>, dwFlags=<unavailable>, wszAppDomainManagerAssemblyName=<unavailable>, wszAppDomainManagerTypeName=<unavailable>, nProperties=8, pPropertyNames=0x0000aaaaaaaf7650, pPropertyValues=0x0000aaaaaaaf7d10, pAppDomainID=0x0000ffffffffe8cc) at corhost.cpp:603:15
    frame #5: 0x0000fffff7359b24 libcoreclr.so`::coreclr_initialize(exePath=<unavailable>, appDomainFriendlyName=<unavailable>, propertyCount=8, propertyKeys=<unavailable>, propertyValues=<unavailable>, hostHandle=0x0000ffffffffe8d0, domainId=0x0000ffffffffe8cc) at exports.cpp:254:16
    frame #6: 0x0000fffff79bce30 libhostpolicy.so`___lldb_unnamed_symbol783 + 552
    frame #7: 0x0000fffff79d7e98 libhostpolicy.so`___lldb_unnamed_symbol1897 + 508
    frame #8: 0x0000fffff79d7874 libhostpolicy.so`corehost_main + 400
    frame #9: 0x0000fffff7a6f590 libhostfxr.so`___lldb_unnamed_symbol1353 + 672
    frame #10: 0x0000fffff7a6ed9c libhostfxr.so`___lldb_unnamed_symbol1349 + 280
    frame #11: 0x0000fffff7a6c37c libhostfxr.so`___lldb_unnamed_symbol1334 + 436
    frame #12: 0x0000fffff7a6b798 libhostfxr.so`___lldb_unnamed_symbol1331 + 372
    frame #13: 0x0000fffff7a653e0 libhostfxr.so`hostfxr_main_startupinfo + 192
    frame #14: 0x0000aaaaaaab0148 dotnet`___lldb_unnamed_symbol589 + 1196
    frame #15: 0x0000aaaaaaab0454 dotnet`___lldb_unnamed_symbol590 + 176
```

**Comment by carlossanlop** [comment]
What is `dotnet.sh`  in your command `/dotnet.sh test src/libraries/System.Formats.Asn1/tests`?

Can you try this, see if it works?:

```
cd path/to/runtime
.dotnet/dotnet build /t:test src/libraries/System.Formats.Asn1/tests/System.Formats.Asn1.Tests.csproj
```

**Comment by vcsjones** [comment]
> What is `dotnet.sh` in your command `/dotnet.sh test src/libraries/System.Formats.Asn1/tests`?


It's the `dotnet.sh` script [in the root of the repository](https://github.com/dotnet/runtime/blob/main/dotnet.sh).

> Can you try this, see if it works?

Same result with different text output. It results in a segmentation fault.

```
----- end Fri Sep 9 05:42:19 PM EDT 2022 ----- exit code 139 ----------------------------------------------------------
exit code 139 means SIGSEGV Illegal memory access. Deref invalid pointer, overrunning buffer, stack overflow etc. Core dumped.
```

**Comment by ViktorHofer** [comment]
@jkotas any idea? Presumably we would need a core dump to understand where the SIGSEGV is coming from.

**Comment by vcsjones** [comment]
> You can try building with debug runtime. If it does not repro with debug runtime, it suggests that the problem is caused by a C++ compiler optimizations bug.

That was a good suggestion. I cannot reproduce it with a runtime debug build.

I can still reproduce it with a runtime release build as of 3e8a5b1852af3a016c442ea5e303afd762216b41. 

Here is the attached [core-dump.tar.gz](https://github.com/dotnet/runtime/files/9801786/core-dump.tar.gz).

Some other potentially helpful info:

`clang --version`:

Ubuntu clang version 14.0.0-1ubuntu1
Target: aarch64-unknown-linux-gnu
Thread model: posix
InstalledDir: /usr/bin

`lsb_release -a`:

Distributor ID:	Ubuntu
Description:	Ubuntu 22.04 LTS
Release:	22.04
Codename:	jammy

I will also point out that this is in an ARM64 Linux VM running running in Parallels on an M1 Mac.

**Comment by janvorli** [comment]
@vcsjones what is the page size on your Linux? The crashing instruction seems to indicate the page size is 128kB.

**Comment by janvorli** [comment]
This looks like a clang assembler bug. There are only variants of the stub code for 4kB, 8kB, 16kB, 32kB and 64kB pages. The instruction dumped by @jkotas above should reference data 1 memory page after the current code point, but it references data 128kB far away. 
Let me take a look at the official build binaries.

**Comment by vcsjones** [comment]
> Let me take a look at the official build binaries.

I was able to run .NET 7 RC2 and .NET 8 nightlies from [dotnet/installer](https://github.com/dotnet/installer). I presume they are built with a different version of clang.

**Comment by janvorli** [comment]
@vcsjones I've tried to cross build the runtime using clang14 on Ubuntu 22.04 x64 (targeting arm64) and the generated code for the problematic helper was correct. It was using exactly the same version of clang as your build. The only difference being that your build was executed on arm64 while my one on x64 cross targeting arm64. 
I am going to try to build it on arm64 Linux in docker to see.

**Comment by janvorli** [comment]
@vcsjones I have M1 Mac, but don't have access to Parallels. But could you please build the runtime as release with an extra option as described below and share the produced libcoreclr.so with me? You can find it in `artifacts/bin/coreclr/Linux.arm64.Release` subfolder after the build completes.
```sh
./build.sh -rc release -s clr+libs -keepnativesymbols
```

I have tried to build the runtime on my arm64 Linux box in docker with Ubuntu 22.04 image and the generated code was correct.

**Comment by vcsjones** [comment]
> ```shell
> ./build.sh -rc release -s clr+libs -keepnativesymbols
> ```

Using that exact command produced:

> No value for keepNativeSymbols is supplied. See help (--help) for supported values.

So I took a gamble and used `-keepnativesymbols true`. If I did something wrong here, please let me know.

Here is the resulting libcoreclr.so: [libcoreclr.so.tar.gz](https://github.com/dotnet/runtime/files/9804760/libcoreclr.so.tar.gz)

**Comment by janvorli** [comment]
The file has really incorrect offsets in the LDR instructions in all the helpers. I am not sure why I cannot repro this build issue in the Ubuntu 22.04 arm64 docker container, it should use the same compiler (and according to what's logged in the binary, the clang version is exactly the same). The only difference might be the linker. I've tried to link it both with the gnu linker and lld and my results were always valid. But could you please try to install `lld-14` package and make a clean build to see if it helps?

**Comment by janvorli** [comment]
@vcsjones I was able to repro it - it is really caused by the gnu linker. When I have tried to link with it before, I've somehow incorrectly uninstalled the lld-14, so it was still being used and I haven't noticed that. So please install the `lld-14` and you should be good.

**Comment by jkotas** [comment]
https://github.com/dotnet/runtime/blob/main/docs/workflow/requirements/linux-requirements.md says to install `lld`. Does this mean that it is necessary to install `lld-14` explicitly?

**Comment by vcsjones** [comment]
> @vcsjones I was able to repro it - it is really caused by the gnu linker. 

Wow. That fixed it.

> https://github.com/dotnet/runtime/blob/main/docs/workflow/requirements/linux-requirements.md says to install `lld`. Does this mean that it is necessary to install `lld-14` explicitly?

I don't think so. Doing `apt-get install lld` did install lld-14. So I just completely missed lld as a dependency.

I apparently just made a mistake following the documentation. Apologies for taking everyone's time here, thank you both.

If the GNU linker doesn't work (is it expected to?) perhaps it's worth failing the build somewhere around here

https://github.com/dotnet/runtime/blob/6ee57c0d1a8ef6ea6a5886b3367f04702fa07e58/eng/native/configuretools.cmake#L85

**Comment by janvorli** [comment]
@vcsjones I have just figured out how to modify the asm source for those helpers so that the linker doesn't play any role here. I've found that the current form has emitted relocations into the object file and the GNU linker then converted those to compile time constants. That's where the GNU linker got it wrong. The modification I have figured results in no relocations in the .obj files, so the linker doesn't do anything with the offsets.
